haystack/database/elasticsearch.py

```diff
@@ -103,24 +105,19 @@ def get_all_documents(self):
     def query(
         self,
         query: str,
+        filters: dict,
         top_k: int = 10,
-        candidate_doc_ids: [str] = None,
-        direct_filters: dict = None,
         custom_query: str = None,
     ) -> [Document]:
-        # TODO:
-        # for now: we keep the current structure of candidate_doc_ids for compatibility with SQL documentstores
-        # midterm: get rid of it and do filtering with tags directly in this query
 
-        # if a custom search query is provided then use it
-        if custom_query:
-            body = {
-                "size": top_k,
-                "query": {
-                    "bool": custom_query
-                }
-            }
-        # else use standard search query for provided search fields
+        if custom_query:  # if a custom search query is provided then use it
+            template = Template(custom_query)
```

Comment:
How fail-safe is this (e.g. a custom query template not using "question")? 
Please also add a small comment here on what is happening ("Exchanging placeholders in the custom_query template with actual values of the current query (e.g. ${question} => str(query) )

Reply:
To be fail-safe, the string must be valid Elasticsearch DSL with `$question` placeholder present. An assertion for at least one question placeholder is added in the Retriever initialization.